File: algorithms/js/letterCombinationsOfAPhoneNumber/letterCombinationsOfAPhoneNumber.js
// Source : https://leetcode-cn.com/problems/letter-combinations-of-a-phone-number/
// Author : Tian Zhuo
// Date   : 2018-11-28

/**
 * @param {string} digits
 * @return {string[]}
 */
var letterCombinations = function(digits) {
    const map = ["", "", "abc", "def", "ghi", "jkl", "mno", "pqrs", "tuv", "wxyz"];
    digits.replace('1', '')
    let res = []
    for (var i = 0; i < digits.length; i++) {
        let j = 0;
        let letters = map[digits[i]]
        if(i == 0) {
            res = letters.split('')
        }else{
            let temp = []
            res.map(e => {
                let k = 0
                while(k < letters.length) {
                    temp.push(e + letters[k])
                    k++
                }

            })
            res = temp;
        }
    }
    return res
};

/**
 * @param {string} digits
 * @return {string[]}
 */
let letterCombinations = function(digits) {
    if (digits === '') return [];
    let dict = ["", "", "abc", "def", "ghi", "jkl", "mno", "pqrs", "tuv", "wxyz"];
    let res = [''];
    for (let digit of digits) {
        let len = res.length;
        while (len-- > 0) {
            let e = res.shift();
            for (let c of dict[digit]) {
                res.push(e + c);
            }
        }
    }
    return res;
};

File: src/main/java/letterCombinationsOfAPhoneNumber/LetterCombinationsOfAPhoneNumber.java
/*
 * @Source: https://leetcode-cn.com/problems/letter-combinations-of-a-phone-number/
 * @Author: Tian Zhuo
 * @Date: 2018-12-17 12:46:54
 */
package letterCombinationsOfAPhoneNumber;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

public class LetterCombinationsOfAPhoneNumber {
    public List<String> letterCombinations(String digits) {
        if ("".equals(digits))
            return Collections.emptyList();
        String[] dict = { "", "", "abc", "def", "ghi", "jkl", "mno", "pqrs", "tuv", "wxyz" };
        List<String> res = new ArrayList<>();
        res.add("");
        String[] digitsArr = digits.split("");
        for (String digit : digitsArr) {
            int len = res.size();
            while (len-- > 0) {
                String e = res.get(0);
                res.remove(0);
                String digitStr = dict[Integer.parseInt(digit)];
                for (int i = 0; i < digitStr.length(); i++) {
                    res.add(e + digitStr.charAt(i));
                }
            }
        }
        return res;
    }
}